## ***TASK 3: 

# **Import Libraries**

In [10]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

# **Import Dataset**

In [13]:
df = pd.read_csv('car.csv')
df

Car_Name  Year  Selling_Price  Present_Price  Driven_kms Fuel_Type  \
0       ritz  2014           3.35           5.59       27000    Petrol   
1        sx4  2013           4.75           9.54       43000    Diesel   
2       ciaz  2017           7.25           9.85        6900    Petrol   
3    wagon r  2011           2.85           4.15        5200    Petrol   
4      swift  2014           4.60           6.87       42450    Diesel   
..       ...   ...            ...            ...         ...       ...   
296     city  2016           9.50          11.60       33988    Diesel   
297     brio  2015           4.00           5.90       60000    Petrol   
298     city  2009           3.35          11.00       87934    Petrol   
299     city  2017          11.50          12.50        9000    Diesel   
300     brio  2016           5.30           5.90        5464    Petrol   

    Selling_type Transmission  Owner  
0         Dealer       Manual      0  
1         Dealer       Manual      0  
2         Dealer       Manual      0  
3         Dealer       Manual      0  
4         Dealer       Manual      0  
..           ...          ...    ...  
296       Dealer       Manual      0  
297       Dealer       Manual      0  
298       Dealer       Manual      0  
299       Dealer       Manual      0  
300       Dealer       Manual      0  

[301 rows x 9 columns]

# **Data Preprocessing**

In [16]:
df.head()

Car_Name  Year  Selling_Price  Present_Price  Driven_kms Fuel_Type  \
0     ritz  2014           3.35           5.59       27000    Petrol   
1      sx4  2013           4.75           9.54       43000    Diesel   
2     ciaz  2017           7.25           9.85        6900    Petrol   
3  wagon r  2011           2.85           4.15        5200    Petrol   
4    swift  2014           4.60           6.87       42450    Diesel   

  Selling_type Transmission  Owner  
0       Dealer       Manual      0  
1       Dealer       Manual      0  
2       Dealer       Manual      0  
3       Dealer       Manual      0  
4       Dealer       Manual      0

In [18]:
df.tail()

Car_Name  Year  Selling_Price  Present_Price  Driven_kms Fuel_Type  \
296     city  2016           9.50           11.6       33988    Diesel   
297     brio  2015           4.00            5.9       60000    Petrol   
298     city  2009           3.35           11.0       87934    Petrol   
299     city  2017          11.50           12.5        9000    Diesel   
300     brio  2016           5.30            5.9        5464    Petrol   

    Selling_type Transmission  Owner  
296       Dealer       Manual      0  
297       Dealer       Manual      0  
298       Dealer       Manual      0  
299       Dealer       Manual      0  
300       Dealer       Manual      0

In [20]:
df.describe()

Year  Selling_Price  Present_Price     Driven_kms       Owner
count   301.000000     301.000000     301.000000     301.000000  301.000000
mean   2013.627907       4.661296       7.628472   36947.205980    0.043189
std       2.891554       5.082812       8.642584   38886.883882    0.247915
min    2003.000000       0.100000       0.320000     500.000000    0.000000
25%    2012.000000       0.900000       1.200000   15000.000000    0.000000
50%    2014.000000       3.600000       6.400000   32000.000000    0.000000
75%    2016.000000       6.000000       9.900000   48767.000000    0.000000
max    2018.000000      35.000000      92.600000  500000.000000    3.000000

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Car_Name       301 non-null    object 
 1   Year           301 non-null    int64  
 2   Selling_Price  301 non-null    float64
 3   Present_Price  301 non-null    float64
 4   Driven_kms     301 non-null    int64  
 5   Fuel_Type      301 non-null    object 
 6   Selling_type   301 non-null    object 
 7   Transmission   301 non-null    object 
 8   Owner          301 non-null    int64  
dtypes: float64(2), int64(3), object(4)
memory usage: 21.3+ KB


In [24]:
df.shape

(301, 9)

In [26]:
df.duplicated().any()

True

In [28]:
duplicate_values = df.duplicated().sum()
duplicate_values

2

In [30]:
df.drop_duplicates(inplace= True)

In [32]:
null_values = df.isna().sum()
null_values

Car_Name         0
Year             0
Selling_Price    0
Present_Price    0
Driven_kms       0
Fuel_Type        0
Selling_type     0
Transmission     0
Owner            0
dtype: int64

# **Data Analysis**

In [35]:
numerical_columns = ['Year', 'Selling_Price', 'Present_Price', 'Driven_kms', 'Owner']

numerical_df = df[numerical_columns]

corr_matrix = numerical_df.corr()

In [37]:
corr_matrix

Year  Selling_Price  Present_Price  Driven_kms     Owner
Year           1.000000       0.234369      -0.053167   -0.525714 -0.181639
Selling_Price  0.234369       1.000000       0.876305    0.028566 -0.087880
Present_Price -0.053167       0.876305       1.000000    0.205224  0.009948
Driven_kms    -0.525714       0.028566       0.205224    1.000000  0.089367
Owner         -0.181639      -0.087880       0.009948    0.089367  1.000000

# **Feature Engineering¶**

In [40]:
df['Car_Age'] = 2024 - df['Year']
df.head()

Car_Name  Year  Selling_Price  Present_Price  Driven_kms Fuel_Type  \
0     ritz  2014           3.35           5.59       27000    Petrol   
1      sx4  2013           4.75           9.54       43000    Diesel   
2     ciaz  2017           7.25           9.85        6900    Petrol   
3  wagon r  2011           2.85           4.15        5200    Petrol   
4    swift  2014           4.60           6.87       42450    Diesel   

  Selling_type Transmission  Owner  Car_Age  
0       Dealer       Manual      0       10  
1       Dealer       Manual      0       11  
2       Dealer       Manual      0        7  
3       Dealer       Manual      0       13  
4       Dealer       Manual      0       10

# **Data Transformation**

In [43]:
df.dtypes

Car_Name          object
Year               int64
Selling_Price    float64
Present_Price    float64
Driven_kms         int64
Fuel_Type         object
Selling_type      object
Transmission      object
Owner              int64
Car_Age            int64
dtype: object

In [45]:
df['Fuel_Type'].value_counts()

Fuel_Type
Petrol    239
Diesel     58
CNG         2
Name: count, dtype: int64

In [47]:
df['Selling_type'].value_counts()

Selling_type
Dealer        193
Individual    106
Name: count, dtype: int64

In [49]:
df['Transmission'].value_counts()

Transmission
Manual       260
Automatic     39
Name: count, dtype: int64

In [51]:
ordinal_map = {'CNG':2,'Diesel':1,'Petrol':0}
df['Fuel_Type'] = df['Fuel_Type'].map(ordinal_map)
df['Fuel_Type'] = df['Fuel_Type'].astype('int64')

In [53]:

df = pd.get_dummies(df, columns = ['Selling_type','Transmission'])

In [55]:
df.dtypes

Car_Name                    object
Year                         int64
Selling_Price              float64
Present_Price              float64
Driven_kms                   int64
Fuel_Type                    int64
Owner                        int64
Car_Age                      int64
Selling_type_Dealer           bool
Selling_type_Individual       bool
Transmission_Automatic        bool
Transmission_Manual           bool
dtype: object

In [57]:
bool_columns = ['Selling_type_Dealer','Selling_type_Individual', 'Transmission_Automatic', 'Transmission_Manual']

df[bool_columns] = df[bool_columns].astype('int64')

In [59]:
df.head()

Car_Name  Year  Selling_Price  Present_Price  Driven_kms  Fuel_Type  Owner  \
0     ritz  2014           3.35           5.59       27000          0      0   
1      sx4  2013           4.75           9.54       43000          1      0   
2     ciaz  2017           7.25           9.85        6900          0      0   
3  wagon r  2011           2.85           4.15        5200          0      0   
4    swift  2014           4.60           6.87       42450          1      0   

   Car_Age  Selling_type_Dealer  Selling_type_Individual  \
0       10                    1                        0   
1       11                    1                        0   
2        7                    1                        0   
3       13                    1                        0   
4       10                    1                        0   

   Transmission_Automatic  Transmission_Manual  
0                       0                    1  
1                       0                    1  
2                       0                    1  
3                       0                    1  
4                       0                    1

# **Model Building¶**

In [62]:
X = df.drop(['Car_Name','Selling_Price'],axis=1)
y = df['Selling_Price']

In [64]:
X.head()

Year  Present_Price  Driven_kms  Fuel_Type  Owner  Car_Age  \
0  2014           5.59       27000          0      0       10   
1  2013           9.54       43000          1      0       11   
2  2017           9.85        6900          0      0        7   
3  2011           4.15        5200          0      0       13   
4  2014           6.87       42450          1      0       10   

   Selling_type_Dealer  Selling_type_Individual  Transmission_Automatic  \
0                    1                        0                       0   
1                    1                        0                       0   
2                    1                        0                       0   
3                    1                        0                       0   
4                    1                        0                       0   

   Transmission_Manual  
0                    1  
1                    1  
2                    1  
3                    1  
4                    1

In [66]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [67]:
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_absolute_error,mean_squared_error, r2_score

## **Linear Regression Model**

In [71]:
from sklearn.linear_model import LinearRegression


# Build a Linear Regression model
lin_reg = LinearRegression()

# Fit the model on the training data
lin_reg.fit(X_train, y_train)

# Predict using the model on the data
predictions = lin_reg.predict(X_test)

In [113]:
  print("Evaluation Results of Linear Regression Model")
  print('Mean Absolute Error', mean_absolute_error(y_test, predictions))
  print('Mean Squared Error', mean_squared_error(y_test, predictions))
  print('RMSE', np.sqrt(mean_squared_error(y_test, predictions)))
  print('R-squared', r2_score(y_test, predictions))

Evaluation Results of Linear Regression Model
Mean Absolute Error 1.5408119549693882
Mean Squared Error 6.67544410049312
RMSE 2.5836880811144987
R-squared 0.7409934356413206


## **Conclusion**

As we can seen R-squared 0.74 > 0.7 indicates model is fit.<br>
As we can seen below following:  <br>
MAE  = 1.54   <br>
MSE  = 6.67   <br>
RMSE = 2.58   <br>
lower MAE, MSE, RMSE indicate better accuracy of model.